index.php

```diff
@@ -567,30 +568,56 @@ private function getUpdateServerResponse(): array {
 	/**
 	 * Downloads the nextcloud folder to $DATADIR/updater-$instanceid/downloads/$filename
 	 *
+	 * Logs download progress
+	 * Resumes incomplete downloads if possible
+	 * Supports outbound proxy usage
+	 * Logs download statistics upon completion
+	 *
+	 * TODO: Provide download progress in real-time (in both CLI and Web modes)
+	 *
 	 * @throws \Exception
 	 */
 	public function downloadUpdate(): void {
 		$this->silentLog('[info] downloadUpdate()');
 
 		$response = $this->getUpdateServerResponse();
-
-		$storageLocation = $this->getUpdateDirectoryLocation() . '/updater-'.$this->getConfigOptionMandatoryString('instanceid') . '/downloads/';
-		if (file_exists($storageLocation)) {
-			$this->silentLog('[info] storage location exists');
-			$this->recursiveDelete($storageLocation);
-		}
-		$state = mkdir($storageLocation, 0750, true);
-		if ($state === false) {
-			throw new \Exception('Could not mkdir storage location');
-		}
-
 		if (!isset($response['url']) || !is_string($response['url'])) {
 			throw new \Exception('Response from update server is missing url');
 		}
 
-		$fp = fopen($storageLocation . basename($response['url']), 'w+');
+		$storageLocation = $this->getUpdateDirectoryLocation() . '/updater-'.$this->getConfigOptionMandatoryString('instanceid') . '/downloads/';
+		$saveLocation = $storageLocation . basename($response['url']);
+		$this->previousProgress = 0;
+
 		$ch = curl_init($response['url']);
+
+		if (!file_exists($storageLocation)) {
+			$state = mkdir($storageLocation, 0750, true);
+			if ($state === false) {
+				throw new \Exception('Could not mkdir storage location');
+			}
+			$this->silentLog('[info] storage location created');
+		} else {
+			$this->silentLog('[info] storage location already exists');
+			// clean-up leftover extracted content from any prior runs, but leave any downloaded Archives alone
+			if (file_exists($storageLocation . 'nextcloud/')) {
+				$this->silentLog('[info] extracted Archive location exists');
+				$this->recursiveDelete($storageLocation . 'nextcloud/');
+			}
```

Comment:
Could we instead delete any file that is not `$saveLocation` to clean out old downloads?